### ISSUE TITLE
'Decoder reached invalid state' panic from Reqwest crate

### ISSUE THREAD

**Issue by jnicholls** [Nemo157/async-compression#79]
The async version of reqwest uses this crate for asynchronous Gzip decoding. I am receiving random 'Decoder reached invalid state' panics in the decoder's stream. Reading the code, I don't understand how that can happen unless the `poll_next` method is being invoked simultaneously from different threads...? Is that accurate? Or is there another condition that could cause this problem? Thanks in advance!

```
thread 'main' panicked at 'Decoder reached invalid state', /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/async-compression-0.2.0/src/stream/generic/decoder.rs:129:35
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:84
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:61
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1025
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1426
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:65
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:50
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:193
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:210
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:471
  11: bytes::abort
  12: <async_compression::stream::generic::decoder::Decoder<S,D> as futures_core::stream::Stream>::poll_next
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.10.1/<::std::macros::panic macros>:3
  13: <async_compression::stream::GzipDecoder<S> as futures_core::stream::Stream>::poll_next
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/async-compression-0.2.0/src/stream/macros/decoder.rs:68
  14: <reqwest::async_impl::decoder::imp::Decoder as futures_core::stream::Stream>::poll_next
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.10.1/src/async_impl/decoder.rs:176
  15: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/stream/stream/mod.rs:1184
  16: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/stream/stream/next.rs:35
  17: std::future::poll_with_tls_context
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/future.rs:99
  18: reqwest::async_impl::response::Response::chunk::{{closure}}
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.10.1/src/async_impl/response.rs:288
  19: <std::future::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/future.rs:43
  20: std::future::poll_with_tls_context
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/future.rs:99
  21: kube::client::APIClient::request_events::{{closure}}::{{closure}}::{{closure}}
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/kube-0.25.0/src/client/mod.rs:180
  22: <std::future::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/future.rs:43
  23: <futures_util::stream::unfold::Unfold<T,F,Fut> as futures_core::stream::Stream>::poll_next
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/stream/unfold.rs:111
  24: <futures_util::stream::stream::map::Map<St,F> as futures_core::stream::Stream>::poll_next
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/stream/stream/map.rs:92
  25: <futures_util::stream::stream::flatten::Flatten<St> as futures_core::stream::Stream>::poll_next
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/stream/stream/flatten.rs:96
  26: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.4/src/stream.rs:121
  27: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/stream/stream/mod.rs:1184
  28: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/stream/stream/next.rs:35
  29: std::future::poll_with_tls_context
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/future.rs:99
  30: kube::api::reflector::Reflector<K>::single_watch::{{closure}}
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/kube-0.25.0/src/api/reflector.rs:214
  31: <std::future::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/future.rs:43
  32: std::future::poll_with_tls_context
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/future.rs:99
  33: kube::api::reflector::Reflector<K>::poll::{{closure}}
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/kube-0.25.0/src/api/reflector.rs:120
  34: <std::future::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/future.rs:43
  35: std::future::poll_with_tls_context
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/future.rs:99
  36: k8s_aws_nlb_proxy_protocol_operator::poll_services::{{closure}}
             at src/main.rs:67
  37: <std::future::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/future.rs:43
  38: tokio::task::core::Core<T>::poll
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.11/src/task/core.rs:128
  39: tokio::task::harness::Harness<T,S>::poll::{{closure}}::{{closure}}
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.11/src/task/harness.rs:120
  40: core::ops::function::FnOnce::call_once
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libcore/ops/function.rs:232
  41: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/panic.rs:318
  42: std::panicking::try::do_call
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/panicking.rs:292
  43: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  44: std::panicking::try
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/panicking.rs:270
  45: std::panic::catch_unwind
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/panic.rs:394
  46: tokio::task::harness::Harness<T,S>::poll::{{closure}}
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.11/src/task/harness.rs:101
  47: tokio::loom::std::causal_cell::CausalCell<T>::with_mut
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.11/src/loom/std/causal_cell.rs:41
  48: tokio::task::harness::Harness<T,S>::poll
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.11/src/task/harness.rs:100
  49: tokio::task::raw::poll
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.11/src/task/raw.rs:162
  50: tokio::task::raw::RawTask::poll
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.11/src/task/raw.rs:113
  51: time::format::parse_items::try_parse_fmt_string
  52: core::ptr::swap_nonoverlapping_bytes
  53: tokio::runtime::basic_scheduler::BasicScheduler<P>::block_on
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.11/src/runtime/basic_scheduler.rs:142
  54: tokio::runtime::Runtime::block_on::{{closure}}
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.11/src/runtime/mod.rs:411
  55: tokio::runtime::context::enter
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.11/src/runtime/context.rs:72
  56: tokio::runtime::handle::Handle::enter
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.11/src/runtime/handle.rs:34
  57: tokio::runtime::Runtime::block_on
             at /Users/jarred.nicholls/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.11/src/runtime/mod.rs:408
  58: k8s_aws_nlb_proxy_protocol_operator::main
             at src/main.rs:23
  59: std::rt::lang_start::{{closure}}
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/rt.rs:67
  60: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:52
  61: std::panicking::try::do_call
             at src/libstd/panicking.rs:292
  62: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  63: std::panicking::try
             at src/libstd/panicking.rs:270
  64: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  65: std::rt::lang_start_internal
             at src/libstd/rt.rs:51
  66: std::rt::lang_start
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/libstd/rt.rs:67
  67: k8s_aws_nlb_proxy_protocol_operator::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by Nemo157** [comment]
That can occur if the decoder has previously yielded either `Err(_)` or `Ok(None)` and you attempt to poll it again. It would probably be worth distinguishing those two cases and panicking with a more accurate message if either are encountered, there should be minimal cost to adding an extra state for it.

**Comment by Nemo157** [comment]
I think [this comment is wrong](https://github.com/clux/kube-rs/blob/6fc2fc70903f46c69d8239ade8b3266b3a66ef95/src/client/mod.rs#L165) it looks like it will loop around and attempt to read another chunk after encountering an error.

**Comment by jnicholls** [comment]
Great find! I did post an issue in kube-rs and will mention your finding. Thanks!

**Comment by Nemo157** [comment]
I think there is a bug here in that according to the current `Stream` contract we should yield a `None` before we can consider ourselves done. So after yielding an error we should move to the `Done` state, then `Invalid` should only be reached when being polled after `None`.

**Comment by jnicholls** [comment]
Yes, after yielding an error a Stream should produce None and as far as I know, it ought to be able to be polled over and over again and continually yield None when it is done; I don't think panicking is an appropriate response to polling None multiple times.

**Comment by Nemo157** [comment]
I can push a fix for the bug tonight and release 0.3.1, which should be usable with `reqwest 0.10.2+`.

**Comment by jnicholls** [comment]
That is awesome thanks!

**Comment by Nemo157** [comment]
@jnicholls just published 0.3.1, it'd be great if you could see if you still get any error like this with it.

**Comment by jnicholls** [comment]
So far so good! No issues yet. I was able to recreate within a minute or two before, but it’s holding strong with no issues so far :) Thanks!
